File: packages/server/src/lib/auto-reminder.ts
import { prisma } from './prisma';

export async function createAutoFeedingReminder(babyId: string) {
  console.log(`[AutoReminder] Computing prediction for baby ${babyId}`);

  const recentFeedings = await prisma.record.findMany({
    where: { babyId, category: 'feeding' },
    orderBy: { occurredAt: 'desc' },
    take: 30,
  });

  if (recentFeedings.length < 2) return;

  const parsed = recentFeedings.map((r) => ({
    ...r,
    parsedData: JSON.parse(r.data) as Record<string, any>,
  }));

  const bottleRates: number[] = [];
  const breastRates: number[] = [];

  for (let i = 0; i < parsed.length - 1; i++) {
    const current = parsed[i + 1];
    const next = parsed[i];
    const intervalMin = (next.occurredAt.getTime() - current.occurredAt.getTime()) / 60000;

    if (intervalMin <= 0 || intervalMin > 480) continue;

    if (current.type === 'bottle') {
      const ml = current.parsedData.amountMl;
      if (ml && ml > 0) bottleRates.push(intervalMin / ml);
    } else if (current.type === 'breastfeed') {
      const totalMin = (current.parsedData.leftMinutes || 0) + (current.parsedData.rightMinutes || 0);
      if (totalMin > 0) breastRates.push(intervalMin / totalMin);
    }
  }

  const lastFeeding = parsed[0];
  const lastFeedingTime = lastFeeding.occurredAt.getTime();
  let predictedInterval: number | null = null;

  if (lastFeeding.type === 'bottle' && bottleRates.length >= 2) {
    const avgRate = bottleRates.reduce((a, b) => a + b, 0) / bottleRates.length;
    const ml = lastFeeding.parsedData.amountMl || 0;
    predictedInterval = Math.round(avgRate * ml);
  } else if (lastFeeding.type === 'breastfeed' && breastRates.length >= 2) {
    const avgRate = breastRates.reduce((a, b) => a + b, 0) / breastRates.length;
    const totalMin = (lastFeeding.parsedData.leftMinutes || 0) + (lastFeeding.parsedData.rightMinutes || 0);
    predictedInterval = Math.round(avgRate * totalMin);
  }

  if (predictedInterval === null) {
    const intervals: number[] = [];
    for (let i = 0; i < parsed.length - 1; i++) {
      const diff = parsed[i].occurredAt.getTime() - parsed[i + 1].occurredAt.getTime();
      const min = diff / 60000;
      if (min > 0 && min <= 480) intervals.push(min);
    }
    if (intervals.length >= 2) {
      predictedInterval = Math.round(intervals.reduce((a, b) => a + b, 0) / intervals.length);
    }
  }

  if (predictedInterval === null) return;

  const remindAt = new Date(lastFeedingTime + predictedInterval * 60000);

  if (remindAt.getTime() <= Date.now()) {
    console.log(`[AutoReminder] Predicted time ${remindAt.toISOString()} is in the past, skipping`);
    return;
  }

  // Remove old unsent auto reminders for this baby
  await prisma.reminder.deleteMany({
    where: { babyId, source: 'feeding_auto', sent: false },
  });

  const reminder = await prisma.reminder.create({
    data: {
      babyId,
      remindAt,
      source: 'feeding_auto',
      title: '喂奶提醒',
      body: '根据喂养规律，宝宝预计需要喂奶了',
    },
  });

  console.log(`[AutoReminder] Created reminder ${reminder.id} for baby ${babyId}, remindAt=${remindAt.toISOString()} (in ${predictedInterval} min)`);
}

File: backend/autoreminder.go
package main

import (
	"log"
	"math"

	"github.com/google/uuid"
)

// createAutoFeedingReminder 根据近期喂养规律预测下次喂奶时间并创建自动提醒。
func createAutoFeedingReminder(babyID string) {
	log.Printf("[AutoReminder] Computing prediction for baby %s", babyID)

	parsed, err := loadRecentFeedings(babyID, 30)
	if err != nil || len(parsed) < 2 {
		return
	}

	var bottleRates, breastRates []float64
	for i := 0; i < len(parsed)-1; i++ {
		current := parsed[i+1]
		next := parsed[i]
		intervalMin := float64(next.occurredAt-current.occurredAt) / 60000.0
		if intervalMin <= 0 || intervalMin > 480 {
			continue
		}
		if current.typ == "bottle" {
			ml := numField(current.data, "amountMl")
			if ml > 0 {
				bottleRates = append(bottleRates, intervalMin/ml)
			}
		} else if current.typ == "breastfeed" {
			totalMin := numField(current.data, "leftMinutes") + numField(current.data, "rightMinutes")
			if totalMin > 0 {
				breastRates = append(breastRates, intervalMin/totalMin)
			}
		}
	}

	lastFeeding := parsed[0]
	lastFeedingTime := lastFeeding.occurredAt
	var predictedInterval *int

	if lastFeeding.typ == "bottle" && len(bottleRates) >= 2 {
		v := int(math.Round(avg(bottleRates) * numField(lastFeeding.data, "amountMl")))
		predictedInterval = &v
	} else if lastFeeding.typ == "breastfeed" && len(breastRates) >= 2 {
		totalMin := numField(lastFeeding.data, "leftMinutes") + numField(lastFeeding.data, "rightMinutes")
		v := int(math.Round(avg(breastRates) * totalMin))
		predictedInterval = &v
	}

	if predictedInterval == nil {
		var intervals []float64
		for i := 0; i < len(parsed)-1; i++ {
			diff := float64(parsed[i].occurredAt-parsed[i+1].occurredAt) / 60000.0
			if diff > 0 && diff <= 480 {
				intervals = append(intervals, diff)
			}
		}
		if len(intervals) >= 2 {
			v := int(math.Round(avg(intervals)))
			predictedInterval = &v
		}
	}

	if predictedInterval == nil {
		return
	}

	remindAt := lastFeedingTime + int64(*predictedInterval)*60000
	if remindAt <= int64(nowMillis()) {
		log.Printf("[AutoReminder] Predicted time is in the past, skipping")
		return
	}

	// 删除该宝宝旧的未发送自动提醒
	if _, err := db.Exec(`DELETE FROM "Reminder" WHERE babyId = ? AND source = 'feeding_auto' AND sent = 0`, babyID); err != nil {
		log.Printf("[AutoReminder] Failed to clear old reminders: %v", err)
	}

	id := uuid.NewString()
	if _, err := db.Exec(`INSERT INTO "Reminder" (id, babyId, remindAt, source, title, body, refId, sent, createdAt)
		VALUES (?, ?, ?, 'feeding_auto', ?, ?, NULL, 0, ?)`,
		id, babyID, remindAt, "喂奶提醒", "根据喂养规律，宝宝预计需要喂奶了", int64(nowMillis())); err != nil {
		log.Printf("[AutoReminder] Failed to create reminder: %v", err)
		return
	}

	log.Printf("[AutoReminder] Created reminder %s for baby %s (in %d min)", id, babyID, *predictedInterval)
}
